tensorflow_federated/python/common_libs/execution_tracing.py

```diff
@@ -0,0 +1,266 @@
+TracingProvider logs information about the computation in a presentable manner.
+Note that it does not attempt to correlate asynchronous calls, and so using
+this in conjunction with a ThreadDelegatingExecutor will not preserve logging
+order as it otherwise would.
+"""
+from enum import Enum
+import logging
+import re
+from typing import Any, Dict, Optional, Tuple
+
+from tensorflow_federated.python.common_libs import tracing
+
+
+class TraceFilterMode(Enum):
+  VALUES = 0
+  INTRINSICS = 1
+  ALL = 2
+
+
+def _pattern_union(*patterns):
+  pattern_strs = map(lambda p: p.pattern, patterns)
+  return re.compile('|'.join(pattern_strs))
+
+
+_VALUE_PATTERN = re.compile('^create_([a-z]+)$')
+_INTRINSIC_PATTERN = re.compile('^_compute_intrinsic_([a-z_]+)$')
+
+_TRACE_FILTER_PATTERNS = {
+    TraceFilterMode.VALUES:
+        _VALUE_PATTERN,
+    TraceFilterMode.INTRINSICS:
+        _pattern_union(_VALUE_PATTERN, _INTRINSIC_PATTERN),
+    TraceFilterMode.ALL:
+        None,
+}
+
+DEFAULT_FORMAT_STRATEGY = lambda x: '{} {}'.format(x, type(x))
+
+
+class ExecutionTracingProvider(tracing.TracingProvider):
+  """Traces a subset of functions appearing in the current execution context.
+
+  Note: This TracingProvider does not attempt to correlate log messages across
+  threads, so executor stacks using ThreadDelegatingExecutor will generally lose
+  their nesting structure in the logs.
+
+  This implementation does not require storing additional trace context state,
+  so most methods are no-ops.
+  """
+
+  def __init__(self,
+               trace_filter_mode=TraceFilterMode.INTRINSICS,
+               default_format_strategy=DEFAULT_FORMAT_STRATEGY):
+    assert isinstance(trace_filter_mode, TraceFilterMode)
+
+    # establish what we're tracing
+    self._is_traceable_regex = _TRACE_FILTER_PATTERNS[trace_filter_mode]
+
+    # fill in formatting strategies
+    self._default_strategy = default_format_strategy
+    self._formatting_strategies = dict()
+    self._make_formatting_strategies()
+
+    # indentation management
+    # start at -1 so the first call prints at 0
+    self._current_call_level = -1
+
+  def span(
+      self,
+      scope: str,
+      sub_scope: str,
+      nonce: int,
+      parent_span_yield: Optional[None],
+      fn_args: Optional[Tuple[Any, ...]],
+      fn_kwargs: Optional[Dict[str, Any]],
+      trace_opts: Dict[str, Any],
+  ):
+    assert parent_span_yield is None
+    del parent_span_yield, nonce, trace_opts
+
+    regex = self._is_traceable_regex
+    force_trace = regex is None
+    if not self._is_traceable(sub_scope, regex, override=force_trace):
+      # exit span early
+      yield None
+
+    else:
+      # Pre-fn
+      self._current_call_level += 1
+      self._log_trace('call', scope, sub_scope, fn_args, fn_kwargs)
+      yield None
+
+      # Post-fn
+      self._current_call_level -= 1
+      self._log_trace('retr', scope, sub_scope, fn_args, fn_kwargs)
+
+  @classmethod
```

Comment:
Nit: since this doesn't use the class, I don't think this belongs as a class method. It seems more like a private helper function, such as `_pattern_union` above.